### PR TITLE
OneDeploy Functionality

### PR DESCRIPTION
**Pull request by ghost**
OneDeploy functionality for deploying local app.war, app.jar, startup files, static files, and zipped files as well as through an URL. More error cases need to be added and asynchronous deployments still need to be tested. There is a known bug regarding deployment logs that I am working on fixing.

### CODE REVIEW

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 1]
```diff
@@ -138,6 +140,164 @@ public class PushDeploymentController : ApiController
...
+                    Fetch = LocalAppFetch,
+                    DoFullBuildByDefault = false,
+                    Author = author,
+                    AuthorEmail = authorEmail,
+                    Message = message
+                };
+
+                switch (deployType)
+                {
+                    case "war":
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >"war": [](start = 25, length = 6)

Can you make an exhaustive list of all possibly scenarios that can ever happen (sync/async, error/no error/ etc) when using type=war and share it here as a comment (for now). Mention whether your code currently handles it or not. If not, what is the plan. 

Then we will do a similar exercise for other types. #Closed

Comment (ghost) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: async=false : Complete
aysnc=true : Preliminary testing shows that it works as intended. More testing will have to be done.
clean=true : Complete
clean=false : Doesn't seem to work at the moment, I'll look into where deploymentInfo.CleanupTargetDirectory is referenced and investigate from there
path : Code currently does not support path for type = war. This can be easily changed though.
Errors : HTTP 400 when trying to deploy type=war to a Java SE webapp. If path is allowed for type=war, a HTTP 400 will be returned if the customer tried to specify a filename that does not have the .war extension. We will not be able to ensure that the file being deployed by the customer is a war file. 


---
In reply to: [455295447](https://github.com/shrishrirang/kudu/pull/2#discussion_r455295447) [](ancestors = 455295447)

Comment (shrishrirang) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: Some test cases: 
- POST with missing body
- POST with empty body
- path=invalid Linux path
- invalid Content-type
- invalid zip file (in case of wardeploy)
- multiple onedeploy at a time (Invoke onedeploy. Make it take longer by putting an artificial sleep, either hardcode or based on presence of a file, and then invoke the second and see what happens. Whether it fails with appropriate error message or not when type=war? See if the behavior matches wardeploy.
- what else?

---
In reply to: [455354445](https://github.com/shrishrirang/kudu/pull/2#discussion_r455354445) [](ancestors = 455354445,455295447)

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 2]
```diff
@@ -138,6 +140,164 @@ public class PushDeploymentController : ApiController
...
+            [FromUri] string author = null,
+            [FromUri] string authorEmail = null,
+            [FromUri] string deployer = OneDeploy,
+            [FromUri] string message = DefaultMessage)
+        {
+            using (_tracer.Step("OnePushDeploy"))
+            {
+                var response = Request.CreateResponse();
+                //Checks the "type" query parameter, defaults to "static" if no query parameter is provided 
+               var deployType = Request.RequestUri.ParseQueryString()["type"];
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >ParseQueryString() [](start = 51, length = 18)

Parsing is expensive. Do it just once. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 3]
```diff
@@ -138,6 +140,164 @@ public class PushDeploymentController : ApiController
...
+                if (string.IsNullOrWhiteSpace(deployType))
+                {
+                    deployType = "static";
+                }
+                var website_framework = (System.Environment.GetEnvironmentVariable("website_framework".ToLower())).ToLower();
+                //Checks the "path" queary parameter, path root at /wwwroot TBD if need to change to /home
+                string path = Request.RequestUri.ParseQueryString()["path"];
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >path [](start = 69, length = 4)

Refer [https://github.com/shrishrirang/kudu/blob/master/Common/Constants.cs](https://github.com/shrishrirang/kudu/blob/master/Common/Constants.cs) for how to define constant strings so that they are configurable in one place #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 4]
```diff
@@ -138,6 +140,164 @@ public class PushDeploymentController : ApiController
...
+                //Checks the "clean" query parameter, defauts to False if no query parameter is provided
+                var cleanQuery = Request.RequestUri.ParseQueryString()["clean"];
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >ParseQueryString [](start = 52, length = 16)

Parsing is expensive. Do it just once. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 1 above)
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >"war" [](start = 25, length = 5)

Use const string. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 5]
```diff
@@ -354,6 +516,35 @@ bool isFunctionJson(string fullName)
...
+                using (tracer.Step(string.Format("Renaming extractTargetDirectory({0}) to tempDirectory({1})", targetInfo.FullName, moveTarget)))
+                {
+                    targetInfo.MoveTo(moveTarget);
+                }
+            }          
+            var tempFilePath = Path.Combine(tempDirPath, deploymentInfo.FileName);
+            Directory.CreateDirectory(tempDirPath);
+            using (tracer.Step("Saving request content to {0}", tempFilePath))
+            {
+                await content.CopyToAsync(tempFilePath, tracer);
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >await content.CopyToAsync(tempFilePath, tracer); [](start = 16, length = 48)

So app.war will be copied twice, once to a temp location and then to final destination. What are the pros and cons of this approach? #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 6]
```diff
@@ -354,6 +516,35 @@ bool isFunctionJson(string fullName)
...
+                var cleanTask = Task.Run(() => DeleteFilesAndDirsExcept(tempFilePath, tempDirPath, tracer));
+                await Task.WhenAll(cleanTask);
+            }
+            CommitRepo(repository, zipDeploymentInfo);
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >CommitRepo [](start = 12, length = 10)

Is this needed? #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 7]
```diff
@@ -138,6 +140,164 @@ public class PushDeploymentController : ApiController
...
+                bool isCleanDeployment = Convert.ToBoolean(cleanQuery);
+                //Checks the "async" query parameter, defaults to false if no query parameter is provided
+                var asyncQuery = Request.RequestUri.ParseQueryString()["async"];
+                isAsync = Convert.ToBoolean(asyncQuery);
+
+                var deploymentInfo = new ZipDeploymentInfo(_environment, _traceFactory)
+                {
+                    AllowDeploymentWhileScmDisabled = true,
+                    Deployer = deployer,
+                    WatchedFilePath = Path.Combine("WEB-INF", "web.xml"),
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >WatchedFilePath = Path.Combine("WEB-INF", "web.xml"), [](start = 20, length = 53)

This is meant as a way to trigger a restart by touching the watched file. Not needed as you have your own way of triggering a restart. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 8]
```diff
@@ -138,6 +140,164 @@ public class PushDeploymentController : ApiController
...
+                //Checks the "path" queary parameter, path root at /wwwroot TBD if need to change to /home
+                string path = Request.RequestUri.ParseQueryString()["path"];
+                //Checks the "clean" query parameter, defauts to False if no query parameter is provided
+                var cleanQuery = Request.RequestUri.ParseQueryString()["clean"];
+                bool isCleanDeployment = Convert.ToBoolean(cleanQuery);
+                //Checks the "async" query parameter, defaults to false if no query parameter is provided
+                var asyncQuery = Request.RequestUri.ParseQueryString()["async"];
+                isAsync = Convert.ToBoolean(asyncQuery);
+
+                var deploymentInfo = new ZipDeploymentInfo(_environment, _traceFactory)
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >ZipDeploymentInfo [](start = 41, length = 17)

Can you review the correctness of all params in the context of the new API? #Closed

**Comment by shrishrirang** [comment]
>         private static void TryTouchWatchedFile(DeploymentContext context, DeploymentInfoBase deploymentInfo)

Can you confirm we are not touching any watched file? We don't need it as we will be using the restart API #Closed

---
Refers to: Kudu.Core/Deployment/DeploymentManager.cs:880 in 41a9796. [](commit_id = 41a979672c25e9cedfd5519cf252e9426da09fd6, deletion_comment = False)

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 9]
```diff
@@ -138,6 +140,164 @@ public class PushDeploymentController : ApiController
...
+                        if (string.IsNullOrWhiteSpace(path))
+                        {
+                            response.StatusCode = HttpStatusCode.BadRequest;
+                            response.Content = new StringContent("Path must be defined for static file deployments");
+                            return response;
+                        }
+                        else
+                        {
+                            deploymentInfo.FileName = Path.GetFileName(path);
+                            deploymentInfo.TargetPath = Path.GetDirectoryName(path);
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: Can you add a comment (in code) about how targetpath is used and how filename is used?

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 5 above)
Comment (dannysongg) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: Performance is definitely the biggest con of this approach. I primarily chose this approach as I did not think it would have been feasible to completely redesign the deployment pipeline in the time that I had. I am sure there are other pros and cons and I would love to discuss them if you have time. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 6 above)
Comment (dannysongg) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: According to comments left in the code, "the deployment pipeline expects to see an least 1 commit in the IRepository." OneDeploy follows the existing deployment pipeline very closely after the files are copied to the temp repository so I assumed it would be needed. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 1 above)
Comment (shrishrirang) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: When you get a chance to check these scenarios, let's capture the summary.

---
In reply to: [456011982](https://github.com/shrishrirang/kudu/pull/2#discussion_r456011982) [](ancestors = 456011982,455354445,455295447)

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 6 above)
Comment (shrishrirang) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: Add a brief comment explaining this in code.

// Need to commit because .. 
CommitRepo(repository, zipDeploymentInfo);

---
In reply to: [456093988](https://github.com/shrishrirang/kudu/pull/2#discussion_r456093988) [](ancestors = 456093988)

File: Kudu.Core/Deployment/DeploymentManager.cs [hunk 10]
```diff
@@ -693,9 +693,15 @@ private DeployResult GetResult(string id, string activeDeploymentId, bool isDepl
...
+                            await PostDeploymentHelper.RestartMainSiteAsync(_environment.RequestId, new PostDeploymentTraceListener(tracer, logger));
+                        }
+                        else
+                        {
+                            await RestartMainSiteIfNeeded(tracer, logger);
+                        }
+            
                         await PostDeploymentHelper.SyncFunctionsTriggers(_environment.RequestId, new PostDeploymentTraceListener(tracer, logger), deploymentInfo?.SyncFunctionsTriggersPath);
 
                         if (!_settings.RunFromZip() && _settings.TouchWatchedFileAfterDeployment())
```
Comment (shrishrirang) [inline Kudu.Core/Deployment/DeploymentManager.cs]: >TouchWatchedFileAfterDeployment [](start = 65, length = 31)

TouchWatchedFileAfterDeployment() will return true (by default), so TryTouchWatchedFile() will be called. We don't want that as we will be using the RestartMainSiteAsync() to perform a restart.  So to solve that problem, add a new property deploymentInfo.DisableTouchingWatchedFile = false and change the following:
if (!_settings.RunFromZip() && _settings.TouchWatchedFileAfterDeployment())
to:
if (!deploymentInfo.disableTouchingWatchedFile && !_settings.RunFromZip() && _settings.TouchWatchedFileAfterDeployment()) #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 9 above)
Comment (shrishrirang) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: When to use FileName versus when to use TargetPath is not clear. Can you share some details?

---
In reply to: [456028286](https://github.com/shrishrirang/kudu/pull/2#discussion_r456028286) [](ancestors = 456028286)

File: Kudu.Contracts/Deployment/DeploymentInfoBase.cs [hunk 11]
```diff
@@ -68,5 +68,9 @@ public bool IsValid()
         // won't update until after a process restart. Therefore, we copy the needed
         // files into a separate folders and run sync triggers from there.
         public string SyncFunctionsTriggersPath { get; set; } = null;
+
+        public string FileName { get; set; }
+
+        public bool DisableTouchingWatchedFile { get; set;}
```
Comment (shrishrirang) [inline Kudu.Contracts/Deployment/DeploymentInfoBase.cs]: >DisableTouchingWatchedFile [](start = 20, length = 26)

Took a re-look at how this will be used. A better choice of name would be WatchedFileEnabled. That's because the use-case is to check for !DisableTouching (negation of a negation). Instead rename this WatchedFileEnabled and set it to true in the constructor (at the beginning of this file) #Closed

File: Kudu.Core/Deployment/DeploymentManager.cs (hunk 10 above)
Comment (shrishrirang) [reply Kudu.Core/Deployment/DeploymentManager.cs]: if (deploymentInfo.WatchFileEnabled && !_settings.RunFromZip() && _settings.TouchWatchedFileAfterDeployment())

---
In reply to: [456719342](https://github.com/shrishrirang/kudu/pull/2#discussion_r456719342) [](ancestors = 456719342)

**Comment by dannysongg** [comment]
Tested Scenarios:
POST with empty body - no deployment happens. Returns warning
POST with missing body - no deployment happens. Returns error
Invalid Path - deployment fails. Returns error.

Untested Scenarios:
Multiple OneDeploys at once
Invalid content-type
 #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 12]
```diff
@@ -152,15 +154,19 @@ public class PushDeploymentController : ApiController
             using (_tracer.Step("OnePushDeploy"))
             {
                 var response = Request.CreateResponse();
+
                 //Parses all query parameters into queryCollection
                 var queryCollection = Request.RequestUri.ParseQueryString();
                 var deployType = queryCollection[Constants.Type];
-                var website_framework = System.Environment.GetEnvironmentVariable(Constants.Framework).ToLower();
+                //var website_framework = System.Environment.GetEnvironmentVariable(Constants.Framework).ToLower();
+                var website_framework = "tomcat";
```
Comment (dannysongg) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: I am aware that this should not be here, it was left over from testing. It will be removed for the next revision 
 #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 6 above)
Comment (shrishrirang) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: Please add a comment in the code explaining why we need to commit.

---
In reply to: [456717046](https://github.com/shrishrirang/kudu/pull/2#discussion_r456717046) [](ancestors = 456717046,456093988)

**Comment by shrishrirang** [comment]
>         private static void TryTouchWatchedFile(DeploymentContext context, DeploymentInfoBase deploymentInfo)

From your latest commit it seems this won't be called anymore. Update the comment so that we can close the discussion.

---
In reply to: [659624949](https://github.com/shrishrirang/kudu/pull/2#issuecomment-659624949) [](ancestors = 659624949)

---
Refers to: Kudu.Core/Deployment/DeploymentManager.cs:880 in 41a9796. [](commit_id = 41a979672c25e9cedfd5519cf252e9426da09fd6, deletion_comment = False)

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 9 above)
Comment (shrishrirang) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: Can you share details and add it to the code as a comment?

---
In reply to: [456719509](https://github.com/shrishrirang/kudu/pull/2#discussion_r456719509) [](ancestors = 456719509,456028286)

**Comment by shrishrirang** [comment]
Can you check the untested scenarios? They are important scenarios.

---
In reply to: [661257049](https://github.com/shrishrirang/kudu/pull/2#issuecomment-661257049) [](ancestors = 661257049)

**Comment by shrishrirang** [comment]
What about these? This should be listed in the list of important scenarios:

- invalid zip file (in case of wardeploy)
- what else?

Also, the list of tested scenarios does not look like an exhaustive list. Can you carefully review the list again to make sure nothing is missed? Also, capture the status of the testing (tested/not-tested) in the list. Pretty much everything should ideally be tested.

---
In reply to: [661272235](https://github.com/shrishrirang/kudu/pull/2#issuecomment-661272235) [](ancestors = 661272235,661257049)

**Comment by dannysongg** [comment]
Tested Scenarios:
POST with empty body - no deployment happens. Returns warning ( Couldn't read data from file <file-path>, this makes an empty POST)
POST with missing body - no deployment happens. Returns error (The request must be chunked or have a content length)
Provide invalid Path - deployment fails. Returns error (
Multiple OneDeploys at once - Second deployment fails and returns error (There is a deployment currently in progress. Please try again when it completes)
Invalid content-type- deployment fails and returns error (Unable to translate bytes from specified code page to Unicode)
Invalid Zip File - deployment fails and returns error (Central Directory corrupt)
Deploy war to non-Tomcat app - Deployment fails and returns error (war files cannot be deployed to <runtime-stack> webapps)
Deploy jar to non-JavaSE app - Deployment fails and returns error (jar files cannot be deployed to <runtime-stack> webapps)
Provide no path to static deployment -  Deployment fails and returns error (Path must be defined for static file deployments)
Provide no type parameter - Deployment fails and returns error (Deployment type not provided)
Provide invalid type parameter - Deployment fails and returns error (Deployment type not valid)
Provide invalid site URL - Deployment fails and returns error (Could not resolve host)
Provide invalid file URL - Deployment fails and returns error (404 not found)
Provide invalid query parameter - Invalid query parameters get ignored #Closed

**Comment by shrishrirang** [comment]
Nice!

---
In reply to: [661314089](https://github.com/shrishrirang/kudu/pull/2#issuecomment-661314089) [](ancestors = 661314089)

File: Common/Constants.cs [hunk 13]
```diff
@@ -136,5 +136,27 @@ public static TimeSpan MaxAllowedExecutionTime
...
+        public const string Zip = "zip";
+
+        //OneDeploy Query Parameters
+        public const string Type = "type";
+        public const string FilePath = "path";
+        public const string Clean = "clean";
+        public const string Async = "async";
+
+        //OndeDeploy Webapp Frameworks 
+        public const string Framework = "website_framework";
```
Comment (shrishrirang) [inline Common/Constants.cs]: >website_framework [](start = 41, length = 17)

Caps - WEBSITE_FRAMEWORK. Case-sensitive in Linux. #Closed

File: Kudu.Contracts/Deployment/DeploymentInfoBase.cs [hunk 14]
```diff
@@ -68,5 +69,9 @@ public bool IsValid()
         // won't update until after a process restart. Therefore, we copy the needed
         // files into a separate folders and run sync triggers from there.
         public string SyncFunctionsTriggersPath { get; set; } = null;
+
+        public string FileName { get; set; }
```
Comment (shrishrirang) [inline Kudu.Contracts/Deployment/DeploymentInfoBase.cs]: >FileName [](start = 22, length = 8)

Move this closer to "TargetPath". Also, add comments in both so that it is clear what the difference is and why we need both. #Closed

File: Kudu.Core/Deployment/DeploymentManager.cs [hunk 15]
```diff
@@ -291,12 +293,10 @@ public async Task RestartMainSiteIfNeeded(ITracer tracer, ILogger logger)
             {
                 return;
             }
-
             if (_settings.RecylePreviewEnabled())
```
Comment (shrishrirang) [inline Kudu.Core/Deployment/DeploymentManager.cs]: >if  [](start = 12, length = 3)

Not a good idea to introduce diffs unless it is intentional. This increases chances of merge conflicts. #Closed

File: Kudu.Core/Deployment/DeploymentManager.cs [hunk 16]
```diff
@@ -693,12 +693,18 @@ private DeployResult GetResult(string id, string activeDeploymentId, bool isDepl
                     {
                         await builder.Build(context);
                         builder.PostBuild(context);
-
-                        await RestartMainSiteIfNeeded(tracer, logger);
-
+                        if(deploymentInfo.Deployer == "OneDeploy")
```
Comment (shrishrirang) [inline Kudu.Core/Deployment/DeploymentManager.cs]: >if(deploymentInfo.Deployer == "OneDeploy") [](start = 24, length = 42)

Add a comment that explains why you can't reuse RestartMainSiteAsync.

File: Kudu.Core/Deployment/DeploymentManager.cs [hunk 17]
```diff
@@ -693,12 +693,18 @@ private DeployResult GetResult(string id, string activeDeploymentId, bool isDepl
...
+                        {
+                            await PostDeploymentHelper.RestartMainSiteAsync(_environment.RequestId, new PostDeploymentTraceListener(tracer, logger));
```
Comment (shrishrirang) [inline Kudu.Core/Deployment/DeploymentManager.cs]: >PostDeploymentTraceListener [](start = 104, length = 27)

Does the log file contain all relevant logs?

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 18]
```diff
@@ -12,6 +12,7 @@
 using System.Threading;
 using System.Threading.Tasks;
 using System.Web.Http;
+using System.Web.UI;
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >Web.UI; [](start = 13, length = 7)

Is this needed? #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 19]
```diff
@@ -31,6 +32,7 @@ public class PushDeploymentController : ApiController
     {
         private const string ZipDeploy = "ZipDeploy";
         private const string WarDeploy = "WarDeploy";
+        private const string OneDeploy = "OneDeploy";
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: If this is moved to Constants.cs, you can use it instead of hardcoding "OneDeploy" in other files. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 20]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
             }
         }
 
+        [HttpPost]
+        [HttpPut]
+        public async Task<HttpResponseMessage> OnePushDeploy(
+            [FromUri] bool isAsync = false,
+            [FromUri] string author = null,
+            [FromUri] string authorEmail = null,
+            [FromUri] string deployer = OneDeploy,
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: author/authorEmail are needed as you won't be using it #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 21]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
...
+            using (_tracer.Step("OnePushDeploy"))
+            {
+                var response = Request.CreateResponse();
+
+                //Parses all query parameters into queryCollection
+                var queryCollection = Request.RequestUri.ParseQueryString();
+                var deployType = queryCollection[Constants.Type];
+                var website_framework = System.Environment.GetEnvironmentVariable(Constants.Framework).ToLower();
+                //Checks the "path" queary parameter, path root at /wwwroot TBD if need to change to /home
+                string path = queryCollection[Constants.FilePath];
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >string [](start = 16, length = 6)

var for sake of consistency. Or if you prefer to use the type explicitly, do the same elsewhere. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 22]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
...
+
+                //Checks the "clean" query parameter, defauts to False if no query parameter is provided
+                var cleanQuery = queryCollection[Constants.Clean];
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >cleanQuery [](start = 20, length = 10)

cleanString. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 23]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
...
+            [FromUri] string deployer = OneDeploy,
+            [FromUri] string message = DefaultMessage)
+        {
+            using (_tracer.Step("OnePushDeploy"))
+            {
+                var response = Request.CreateResponse();
+
+                //Parses all query parameters into queryCollection
+                var queryCollection = Request.RequestUri.ParseQueryString();
+                var deployType = queryCollection[Constants.Type];
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >Type [](start = 59, length = 4)

Might want to make type, path, etc a param just like isAsync. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 24]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
...
+                var website_framework = System.Environment.GetEnvironmentVariable(Constants.Framework).ToLower();
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >var website_framework = System.Environment.GetEnvironmentVariable(Constants.Framework).ToLower(); [](start = 16, length = 97)

Grouping all similar statements (queryCollection[...]) and then keeping the GetEnvironmentVariable line separate is good for readability #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 25]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
             }
         }
 
+        [HttpPost]
+        [HttpPut]
+        public async Task<HttpResponseMessage> OnePushDeploy(
+            [FromUri] bool isAsync = false,
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >isAsync [](start = 27, length = 7)

Is this case insensitive? isAsync=fAlSe?

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 26]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
...
+                //Checks the "path" queary parameter, path root at /wwwroot TBD if need to change to /home
+                string path = queryCollection[Constants.FilePath];
+
+                //Checks the "clean" query parameter, defauts to False if no query parameter is provided
+                var cleanQuery = queryCollection[Constants.Clean];
+                bool isCleanDeployment = Convert.ToBoolean(cleanQuery);
+
+                //Checks the "async" query parameter, defaults to false if no query parameter is provided
+                var asyncQuery = queryCollection[Constants.Async];
+                isAsync = Convert.ToBoolean(asyncQuery);
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: The function parameter isAsync already does this for you for free. Is there a reason for doing it again? 
 #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 20 above)
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >deployer [](start = 29, length = 8)

What happens if the client does:
POST /api/onedeploy?type=war&deployer=onedeploytest

Do we want this to be customisable? If not, remove it from the query parameter list. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 27]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
...
+
+                var deploymentInfo = new ZipDeploymentInfo(_environment, _traceFactory)
+                {
+                    AllowDeploymentWhileScmDisabled = true,
+                    Deployer = deployer,
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >deployer [](start = 31, length = 8)

Constants.OneDeploy? #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 28]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
...
+                    IsContinuous = false,
+                    AllowDeferredDeployment = false,
+                    IsReusable = false,
+                    CleanupTargetDirectory = isCleanDeployment,
+                    TargetChangeset = DeploymentManager.CreateTemporaryChangeSet(message: "Deploying from pushed file"),
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >TargetChangeset [](start = 20, length = 15)

How is this used: TargetChangeset?

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 29]
```diff
@@ -138,6 +140,185 @@ public class PushDeploymentController : ApiController
...
+                    CommitId = null,
+                    RepositoryType = RepositoryType.None,
+                    Fetch = LocalAppFetch,
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs:182]: >Fetch = LocalAppFetch [](start = 20, length = 21)

Don't you need to flip this depending on Content-Type to support URL based deployments?

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 30]
```diff
@@ -354,6 +541,35 @@ bool isFunctionJson(string fullName)
             }
         }
 
+        private async Task LocalAppFetch(IRepository repository, DeploymentInfoBase deploymentInfo, string targetBranch, ILogger logger, ITracer tracer)
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >LocalAppFetch [](start = 27, length = 13)

Please add comments wherever the code is not self explanatory. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 31]
```diff
@@ -354,6 +541,35 @@ bool isFunctionJson(string fullName)
...
+        {
+            var zipDeploymentInfo = (ZipDeploymentInfo)deploymentInfo;
+            var content = !string.IsNullOrEmpty(zipDeploymentInfo.ZipURL)
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >ZipURL [](start = 66, length = 6)

Will the URL point to a zip if type=war (app.war) ? #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 32]
```diff
@@ -354,6 +541,35 @@ bool isFunctionJson(string fullName)
...
+                ? await DeploymentHelper.GetZipContentFromURL(zipDeploymentInfo, tracer)
+                : Request.Content;
+            var tempDirPath = repository.RepositoryPath;
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >RepositoryPath [](start = 41, length = 14)

Comment needed explaining what is RepositoryPath #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 31 above)
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >var [](start = 12, length = 3)

Let's define this as:
HttpContent content = ...

instead of:
var content = ...

as the type is not obvious and may confuse the reader to believe that this is a string #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 33]
```diff
@@ -354,6 +541,35 @@ bool isFunctionJson(string fullName)
...
+            var targetInfo = FileSystemHelpers.DirectoryInfoFromDirectoryName(tempDirPath);
+            if (targetInfo.Exists)
+            {
+                var moveTarget = Path.Combine(targetInfo.Parent.FullName, Path.GetRandomFileName());
+                using (tracer.Step(string.Format("Renaming extractTargetDirectory({0}) to tempDirectory({1})", targetInfo.FullName, moveTarget)))
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >"Renaming extractTargetDirectory({0}) to tempDirectory({1})", [](start = 49, length = 61)

Why do we do this? #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 34]
```diff
@@ -354,6 +541,35 @@ bool isFunctionJson(string fullName)
             }
         }
 
+        private async Task LocalAppFetch(IRepository repository, DeploymentInfoBase deploymentInfo, string targetBranch, ILogger logger, ITracer tracer)
+        {
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >{ [](start = 8, length = 1)

This method is very difficult to read. Group related lines together, separate unrelated lines with a blank line, add relevant comments (explaining what we are doing and why we are doing) wherever the code is not obvious. #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 28 above)
Comment (dannysongg) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: TargetChangeset is used quite extensively and I can't pin point its exact use but it seems like it tracks changes made to the webapp

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 31 above)
Comment (dannysongg) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: No, I began coding this with the mindset of trying to change as little as possible of the existing code. I have recently realized that that mindset has made my code kind of difficult to read.

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 33 above)
Comment (dannysongg) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: We are doing this to move the old content of the temporary directory in a different one to be deleted concurrently along with anything else unnecessary that may be lingering in the folder  #Closed

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 35]
```diff
@@ -354,6 +525,42 @@ bool isFunctionJson(string fullName)
             }
         }
 
+        private async Task LocalAppFetch(IRepository repository, DeploymentInfoBase deploymentInfo, string targetBranch, ILogger logger, ITracer tracer)
```
Comment (shrishrirang) [inline Kudu.Services/Deployment/PushDeploymentController.cs]: >LocalAppFetch [](start = 27, length = 13)

Can you confirm that d:\home\site\deployments\active file is updated by onedeploy for all scenarios? legacy scenarios like wardeploy/zipdeploy and also for new scenarios like app.war, static, etc? That is needed to query the status.

Comment (ghost) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: Yeah, I can confirm that the active file if updated by OneDeploy for all scenarios
